### PR TITLE
The new euler1d app GUI layout, created by PUI

### PR DESCRIPTION
**Pull request by j8xixo12**
I refine euler1d application GUI layout by [PUI](https://github.com/buganini/PUI).

The latest layout contains dockable solver configure window and a plotting area which is a sub-window.

This PR still has a remaining feature that has not been implemented yet, which is the data selection in single figure layout, I will implement it in next PR.

This is the screenshot of latest GUI:
截圖 2024-01-28 晚上10 09 24

### CODE REVIEW

File: modmesh/app/new_euler1d.py
Comment (yungyuc) [inline modmesh/app/new_euler1d.py:1]: Why don't you directly update the existing `euler1d.py`?

**Review by yungyuc — COMMENTED** [review]
We need some discussions for why you choose to use a new app file `new_euler1d.py` instead of updating the existing `euler1d.py`.

**Comment by j8xixo12** [comment]
> We need some discussions for why you choose to use a new app file `new_euler1d.py` instead of updating the existing `euler1d.py`.

Because this UI still has a remaining feature that has not been implemented yet, which is the data selection in single figure layout.

I plan to implement it and using `new_euler1d.py` to instead of `euler1d.py`.
But I just realized how bad this plan is from the reviewer's perspective because replacing one file with another will create significant differences that are hard to review.

I think the new plan is to update the contents of `euler1d.py` with `new_euler1d.py` directly in this PR, and leave the remaining feature for the next PR.

**Comment by yungyuc** [comment]
You are right it is impossible to review the `new_euler1d.py` without the diff.  When you are ready to request for the next review, please make an explicit global comment saying that.  You tend to use many checkins so that I will ignore the noisy notification from them.

Before you request for review, please properly annotate the code (diff/patch).

One more question: is there anything can be done in the old `euler1d.py` that cannot be done in your (current) `new_euler1d.py`?

**Comment by j8xixo12** [comment]
> One more question: is there anything can be done in the old `euler1d.py` that cannot be done in your (current) `new_euler1d.py`?

Old `euler1d.py` allows user to select which data line is plotted in plotting area under single layout mode.
I plan to integrate this into plot area configuration window that allows user to configure data's color, maker...etc and select which data line to be plotted.
截圖 2024-01-31 晚上11 16 12

Old `euler1d.py` also has a feature that allow user to drag plot to change the layout under grid layout mode, but I think this feature can be replaced with plot configuration window, user can arrange layout on it.

**Comment by yungyuc** [comment]
Thanks.  The data line switches are a good feature but not a must-have for analysis.  It can be added back after the new app layout is merged.  And I also agree that drag-n-drop reordering can be achieved by the plot configuration window.

Thus I think replacing the old `euler1d.py` with the `new_euler1d.py` is a right move.  And after it is done, we may further integrate the standalone application into the viewer system.  It will be a great move forward.  I feel excited.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:75]: Removing bind_axis method, assign axis directly in Line [381](https://github.com/solvcon/modmesh/pull/271/files#diff-4ebeb3148c683103bf030637fdf520154a4b12d9b660a82610861aa5250a0be1R381) and [442](https://github.com/solvcon/modmesh/pull/271/files#diff-4ebeb3148c683103bf030637fdf520154a4b12d9b660a82610861aa5250a0be1R442)
Just like assign `ana` and `num`.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:97]: Data lines are used for visualization, there is no need to keep an array copy, therefore using pass by reference instead of passing a copy of the array.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:92]: Removing the native `PySide6` api and using the PUI to instead of it.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:206]: Removing the GUI layout update callback. Once the GUI needs to be refreshed, this part will be handled in the PUI DOM update.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:232]: Ditto.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:290]: Removing drag and drop reordering feature, it would be reimplemented within the configuration window of the plotting area, but the implementation way may not using drag and drop.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:475]: Data line visualization selection will be moved to the configuration window of the plotting area later.
Therefore the checkbox is removed.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:545]: Since the Euler1d app is constructed with dockable widgets and sub-windows, the option of using sub-windows is meaningless.

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:583]: Using the `PUI` to instead of these lousy imperative code.
The GUI layout is implemented in [688](https://github.com/solvcon/modmesh/pull/271/files#diff-4ebeb3148c683103bf030637fdf520154a4b12d9b660a82610861aa5250a0be1R688) and [645](https://github.com/solvcon/modmesh/pull/271/files#diff-4ebeb3148c683103bf030637fdf520154a4b12d9b660a82610861aa5250a0be1R645)

File: modmesh/app/euler1d.py
Comment (j8xixo12) [inline modmesh/app/euler1d.py:515]: The `Controller` class was removed because it can be replaced by a configuration windows, that allows user modified the solver configuration via GUI.

**Comment by yungyuc** [comment]
I would like to try the code but get

```
ModuleNotFoundError: No module named 'PUI'
```

I don't think we should use PUI as an external dependency because it is not stable yet.  @j8xixo12 could you please incorporate it into modmesh?

**Review by yungyuc — CHANGES_REQUESTED** [review]
I would like to try the code but get

```
ModuleNotFoundError: No module named 'PUI'
```

I don't think we should use PUI as an external dependency because it is not stable yet.  @j8xixo12 could you please incorporate it into modmesh?

**Comment by j8xixo12** [comment]
> I would like to try the code but get
> 
> ```
> ModuleNotFoundError: No module named 'PUI'
> ```
> 
> I don't think we should use PUI as an external dependency because it is not stable yet. @j8xixo12 could you please incorporate it into modmesh?

Does incorporating PUI into modmesh mean putting the entire PUI into modmesh, or can we add PUI as a submodule?
I'm thinking, using a submodule allows us to pin the stable commit of PUI for modmesh usage.

**Comment by j8xixo12** [comment]
> > I would like to try the code but get
> > ```
> > ModuleNotFoundError: No module named 'PUI'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I don't think we should use PUI as an external dependency because it is not stable yet. @j8xixo12 could you please incorporate it into modmesh?
> 
> Does incorporating PUI into modmesh mean putting the entire PUI into modmesh, or can we add PUI as a submodule? I'm thinking, using a submodule allows us to pin the stable commit of PUI for modmesh usage.

@buganini Do you have any idea about how to incorporate PUI into modmesh ?

Since PUI is still under development, users may encounter problems with both modmesh and PUI simultaneously. This could result in a poor user experience and hard to debug for developer.

My suggestion is to add PUI as a submodule into modmesh. This way, modmesh can pin the stable commit of PUI for its usage. When modmesh needs additional features from PUI, we can update the submodule to the next stable commit.
With a submodule, users won't need to install PUI manually, and errors like `ModuleNotFoundError: No module named 'PUI'`should not occur again.

@yungyuc could you give me some suggestions, if I missed something important.

**Comment by yungyuc** [comment]
@j8xixo12 I think we can start with git submodule pinning to a specific changeset.  We may need PUI to come up with a release scheme before we can properly use it as a dependency and manage using devenv.  Could you please help bring it to discord to discuss?

**Comment by j8xixo12** [comment]
> @j8xixo12 I think we can start with git submodule pinning to a specific changeset. We may need PUI to come up with a release scheme before we can properly use it as a dependency and manage using devenv. Could you please help bring it to discord to discuss?

Ok, I will bring it to discord and have a discussion with bug.
But I guess PUI needs some time to come out the release scheme, I will pin the latest commit for now.
@yungyuc what do you think ?

**Comment by yungyuc** [comment]
> Ok, I will bring it to discord and have a discussion with bug. But I guess PUI needs some time to come out the release scheme, I will pin the latest commit for now. @yungyuc what do you think ?

Let's pin to a changeset/commit at the moment and work with PUI for a release scheme.  Those are two separate things.

**Comment by j8xixo12** [comment]
@yungyuc I have added PUI as a submodule into modmesh, because PUI is a thirdparty library, therefore I put it into `thirdparty` directory.
It will be preload during modmesh initialization.

File: modmesh/view.py
Comment (j8xixo12) [inline modmesh/view.py]: If user forget to initialize the submodule it will raise a `ImportError` and terminate modmesh, also the error message will remind user to initialize the submodule to install the PUI.

**Review by yungyuc — CHANGES_REQUESTED** [review]
- [x] Add an automation script under `thirdparty/` to document how to pull PUI.
- [x] Add a unit test to exercise importing PUI and make sure it basically works.

File: .gitmodules
Comment (yungyuc) [inline .gitmodules:1]: It's good to put it under `thirdparty/` which is an existing directory.  I suggest to also add an automation script under it to document how to pull PUI.

File: tests/test_view.py
Comment (j8xixo12) [inline tests/test_view.py:51]: Because `PUINode` and `PUIView` are key components in `PUI`, without these two attributes, `PUI` will not work. Therefore, I chose to check if these two attributes are imported properly.

File: tests/test_view.py
Comment (j8xixo12) [inline tests/test_view.py:53]: Make sure PUI use the right backend which is `PySide6`.

File: modmesh/view.py
Comment (j8xixo12) [inline modmesh/view.py]: If `view` is imported from `modmesh` library, that means modmesh is built with Qt. This indicates modmesh has GUI and the PUI needs to be imported otherwise PUI does not need to be imported.

File: CMakeLists.txt
Comment (j8xixo12) [inline CMakeLists.txt:41]: If modmesh is built with Qt, then PUI also needs to be installed, therefore I let PUI to be installed automatically.

File: .gitmodules
Comment (j8xixo12) [reply .gitmodules:1]: I add the `PUI_install.py` to pull and install PUI automatically.

File: modmesh/view.py
Comment (j8xixo12) [inline modmesh/view.py:78]: Check if the `PUI` folder in `thirdparty` contains `PUI` folder or not, if the user does not initialize the submodule, then the `PUI` in `thirdparty` will be an empty folder.

File: thirdparty/PUI_install.py
Comment (j8xixo12) [inline thirdparty/PUI_install.py:7]: Using the same way in `modmesh/view.py` to check if the submodule is initialized or not.

**Comment by j8xixo12** [comment]
- [x] Add an automation script under thirdparty/ to document how to pull PUI.
- [x] Add a unit test to exercise importing PUI and make sure it basically works.

@yungyuc  this PR is ready for review.

File: thirdparty/PUI_install.py
Comment (yungyuc) [inline thirdparty/PUI_install.py:1]: Please use a (bash) shell script.  It baffled me why you choose to use a Python script when the two other install scripts [googletest.sh](https://github.com/solvcon/modmesh/blob/master/thirdparty/googletest.sh) and [metal-cpp.sh](https://github.com/solvcon/modmesh/blob/master/thirdparty/metal-cpp.sh) in the directory exclusively uses bash.

Using bash allows the maintainers to manage the thirdparty without needing to install Python (3).  It's the same rationale of using bash for [devenv](https://github.com/solvcon/devenv).  On many server farms (or supercomputers) there may not have up-to-date Python or other dependencies installed.

File: tests/test_view.py
Comment (yungyuc) [reply tests/test_view.py:51]: This is sweet.

File: tests/test_view.py
Comment (yungyuc) [reply tests/test_view.py:53]: It's clear documentation.  Thanks.

**Review by yungyuc — CHANGES_REQUESTED** [review]
Thanks, @j8xixo12 .  Please fix the following:

- [x] Use bash for the install script of PUI.

File: thirdparty/PUI_install.py
Comment (j8xixo12) [reply thirdparty/PUI_install.py:1]: The reason for using a Python script is that bash can't run on Windows. However, modmesh claims to support Windows. Also, our CI has a Windows version, and Python as a dependency of modmesh, user cannot run modmesh without Python. Therefore, I choose to use Python script to implement the install script.

Comment (yungyuc) [reply thirdparty/PUI_install.py:1]: I didn't notice that you made it in Python for Windows, which is reasonable.  It's good to use Python then.  Thank you for taking care of it.

File: modmesh/view.py
Comment (yungyuc) [inline modmesh/view.py:67]: We should make the consumption of PUI more systematic in the future, so that we do not need to hard-code import path like here.  While we can live with it now, @j8xixo12 could you please create an issue to track the improvement for this?

**Review by yungyuc — APPROVED** [review]
Thanks @j8xixo12 for the great improvement.

**Comment by yungyuc** [comment]
@j8xixo12 , I got it working once pulling and rebuilding out of the box.  Fantastic!

image

**Comment by yungyuc** [comment]
This PR fixes #212

**Pull request merged**
